### ISSUE TITLE
 FLAC+CUE files  doesn't split to individual FLAC files

### ISSUE THREAD

**Issue by kangbeicai** [andresch/trackfs#4]
trackfs is run in docker and the directory is correctly mounted
but the   FLAC+CUE files  doesn't split to individual FLAC files, all the files are the same with that in the source directory 
![1](https://user-images.githubusercontent.com/57837294/105486216-1a1a5200-5ce9-11eb-8ae0-3c928774098f.png)

**Comment by andresch** [comment]
Hi kangbeicai,

Thanks for giving trackfs a try.

From what I can see in the directory listing above you don't have a audio file in flac encoding with an embedded cue-sheet (flac+cue), but instead a wav audio file with an accompanying cue-sheet file. 

Currently trackfs only works with flac+cue files. 

You might want to have a look into [ffmpgeg](https://github.com/nschlia/ffmpegfs/tree/VERSION_2.3). ffmpegfs has picked up the idea of cue-support but supports a variety of audio and video formats and also accompannying cue-sheets. This is currently still in development, but already working and but might be worth a try.

Best,
Andreas

**Comment by kangbeicai** [comment]
Thanks for your reply.
I tried again to test a flac file but it still doesnt work.
![微信截图_20210123115015](https://user-images.githubusercontent.com/57837294/105568031-6610d980-5d71-11eb-9040-1fca6183f161.png)

**Comment by andresch** [comment]
There are three possible reasons for trackfs not being able to split a flac file:

1. The flac file has no embeedded  cue-sheet in the CUESHEET tag
2. The cuesheet parser failes to parse the cue-sheet
3. Something in the cue-sheet causes the extraction of the track to fail (e.g. invalid time-ranges in the cue sheet)

In all three cases trackfs keeps the original file instead of exposing individual tracks.

I would need your flac file to debug all three cases, but if you have time to help me debugging, we could try to start with the first two topics without running into music-copyright trouble.

For this you would have to extract the cue-sheet from within the running docker container e.g. with

```
docker exec -ti $(docker ps | grep andresch/trackfs | awk '{print $1}') \
  /usr/bin/metaflac --show-tag=CUESHEET "/src/2001 - Secret of the Runes/Therion - 2001 - Secret of the Runes.flac"
```

This should return the content of the embedded cue-sheet. In case this fails, your ripper dipper did not include the cue sheet as tag into the flac file but instead just created an accompanying cue-file. In that case would would have to use some tag-editor to embed the cue-sheet. Otherwise please attach the output to your reply, then I will check if the trackfs cue-sheet parser has any problem with it.

**Comment by kangbeicai** [comment]
I'm not sure if I input the correct command, but it returns nothing. The command you provided here can't work so i use the docker terminal.
![image](https://user-images.githubusercontent.com/57837294/105606617-75297300-5dd5-11eb-85d6-330d53d03a97.png)
![image](https://user-images.githubusercontent.com/57837294/105606574-35628b80-5dd5-11eb-87bf-7545a709ec88.png)

I have uploaded FLAC+CUE to google drive,  Thanks for your help!
https://drive.google.com/drive/folders/1Txlffa63aq4sw15YWNs0WViXZCyZGNhW?usp=sharing

**Comment by andresch** [comment]
Technically this could get implemented, it was just out of scope of this still very early version of trackfs. I wanted to focus on FLAC+CUE files only. 

But there are also other reasons:
1. I found it more robust to solely rely on meta-data of a single flac file, rather than relying on things like "files with a similar name in the same folder" to try to find matching cue sheets. For the same reason I also only use album-art that is embedded in the flac file rather than looking for jpg in the same folder. 

2. Finally properly implementing standalone cue-sheet support, can easily can quite nasty: in theory a cue sheet could refer to multiple different audio files at the same time; and in practice most CD rippers create invalid file-references in the cue-sheet. If a cue-sheet is embedded in a flac file it seems logical to ignore file references in a cue-sheet, but once you use standalone cue sheets as input, this is no longer that logical.

Feel free to create a feature request ticket. I can't promise that I would find time soon, but some basic support should not be to much work either.

Given that the ffmpegfs project that I've mentioned above is currently implementing support for external flac files and would also support cue sheet for WAV files (as in your initial example), if external cue-file support is important/urgent for you, you might want to give this project a try. I've just tested the latest development branch last night and it looks quite promising. I'm sure the maintainer would be keen to get another early tester.

**Comment by kangbeicai** [comment]
I will try ffmpegfs, but it seems that there isn't a docker file release ? I prefer to use docker, otherwise I need to use an additional virtual machine to run it.

**Comment by andresch** [comment]
If you don't mind building the docker image yourself, attached you find the dockerfile I've build for my own tests.

This is not production grade (esp. doesn't run as non-root user), but for testing is should be good enough.

You don't have to prepare anything but building the docker image. The build clones ffmpegfs from github and by default builds the master branch. You can switch to another branch with `--build-arg BRANCH={branch_name}`

E.g. to build the development branch that contains the cue-sheet support I use

```
docker build --build-arg BRANCH=VERSION_2.3 -t andresch/ffmpegfs .
```

The build takes a few minutes, but then you can use the image in a similar way than trackfs.

Before using ffmpegfs there is one thing to do: ffmpegfs somehow has a problem with directly mounting the fuse filesystems on a directly that is already the mountpoint for the docker container. To work around this limitation the docker container expects an empty directory called `ffmpegfs` inside the provided /dst docker mountpoint and uses this as fuse mountpoint. So please make sure that this directory exists in your mountpoint before you run the docker container. If you want a different name for the directly, just modify the corresponding line at the end of the dockerfile and rebuild.

Now you can call:

```
docker run --rm \
    --name=ffmpegfs \
    --device /dev/fuse \
    --cap-add SYS_ADMIN \
    --security-opt apparmor:unconfined \
    -v /path/to/yourmusiclibrary:/src:ro \
    -v /path/to/yourmountpoint:/dst:rshared \
    andresch/ffmpegfs \
  ... whatever ffmpegfs arguments you might want to use
```

e.g. I use ffmpegfs to split all the tracks and decode them into mp3

```
docker run --rm \
    --name=ffmpegfs \
    --device /dev/fuse \
    --cap-add SYS_ADMIN \
    --security-opt apparmor:unconfined \
    -v /path/to/yourmusiclibrary:/src:ro \
    -v /path/to/yourmountpoint:/dst:rshared \
    andresch/ffmpegfs \
    -f --log_stderr --audiobitrate=256K -o allow_other,ro,desttype=mp3
```

ffmpegfs works slightly different than trackfs: ffmpegfs doesn't replace the source-flac file, but instead creates a virtual directly with the same name as the cue-sheet that then contains the individual tracks. 

Let me know if that works for you.

[Dockerfile.zip](https://github.com/andresch/trackfs/files/5862184/Dockerfile.zip)

**Comment by kangbeicai** [comment]
Thanks a lot. I have used your dockerfile to build the dock image.
Some tracks are successfully split into individual tracks but others are not.
I have uploaded one of them to google drive, maybe you can test it ?
https://drive.google.com/drive/folders/1Txlffa63aq4sw15YWNs0WViXZCyZGNhW?usp=sharing

**Comment by andresch** [comment]
In that case this is just a wrong name of the cue-sheet file: ffmpegfs expects that cue-sheet and audio-file have the same basename: e.g. album.flac and album.cue; in your case the name of the cue-sheet is album.flac.cue. After renaming everything works for me.

Exactly the kind of problem I wanted to avoid in trackfs by just using embedded cue-sheets.

**Comment by kangbeicai** [comment]
Yes, that's the reason for that case. 
However, I found ffmpegfs just create an empty directory  in another case, even the cue file is not shown in that directory. I have tried to rename the file name using just English but it didn't work.
https://drive.google.com/drive/folders/1kWHZsXlv3XRjGmRstKnrToW6k7OUxXeB?usp=sharing

**Comment by andresch** [comment]
The log output of ffmpegfs prints:
```
21-01-25 12:09:31 ERROR  : [/src/test/劇場版マクロスF サヨナラノツバサ netabare album the end of”triangle”.wav] Unable to parse cue sheet.
bad character '▒'
bad character '▒'
bad character '▒'
967: syntax error
```
I'm afraid that's a question you have to ask on the ffmpegfs project. I would expect some trouble with utf-8 characters in the cuesheet.

**Comment by andresch** [comment]
I've added basic support for external cue-sheets to trackfs. I do not directly scan the directly for cue files, but if a flac file doesn't has no cue-sheet embedded, then I check if there is a file with the same basename and ".cue"extension and if so I read it. Could not extensively test it (all my flac files are with embedded cues), but at least with the one you provided it works.

I've also added character-encoding-detection to the code,. Would be curious to see if trackfs can process your cue-sheets. 

There is be still some work pending: e.g. currently I only use track-information from the cue-sheet, but expect that album information comes from the flac file. in your example the flac file did not contain any meta-data. But except for some missing tags, I hope it works.

**Comment by kangbeicai** [comment]
sorry for the late reply.  I think trackfs works very well for my music library so far. I will report here if there is any problem.
Also, I'm looking forward to the progress in wav support.
Thinks very much for your work and kindness

**Comment by andresch** [comment]
Just released initial support for wav files. the new version also uses album information from the cue-sheet in case the flac file misses album-related tags. let me know if it works for you.

**Comment by kangbeicai** [comment]
It doesn't work for me ,  maybe it's also because of the utf-8 characters ?
 And there are some problems:
1. if the file is a flac or wav file and there isn't a cue file,  this  file won't be mounted in the mount point.
2. if there is a cue file that has the trouble of utf-8 characters(maybe?), the root directory will be empty. I uploaded such a cue file
[堀江由衣 - イメージマキシシングル フルーツバスケット -風色- song for Yui Horie.zip](https://github.com/andresch/trackfs/files/5925389/-.-.-.song.for.Yui.Horie.zip)
 I think there is nothing about the flac/wav file so I didn't upload flac/wav file.
3. This point is exactly related to point 1, if the files are some subtracks and a cue file, the subtracks also won't be mounted in the mount point. This doesn't matter if the point 1 can be solved.
4. This one is also not important but I found it in the last version. trackfs only recognize the flac file, but not FLAC file,  when I rename the file from FLAC to flac, everything works well.

**Comment by andresch** [comment]
Thanks for testing!

(1) is a bug that I might have introduced as part of the refactoring; I have created a ticket for it (#12) will look into it asap;
(2) will look into that too; might take a bit longer :-)
(4) This is expected; currently I only look for exact matches of ".flac" and ".wav"; will look into making this case insensitive; will create a ticket for that too. for the time being you could try to modify the --extension parameter with an appropriate regular expression.
not sure I understand (3); can you please elaborate?

**Comment by andresch** [comment]
(1) should be fixed; please have a look
The rest might take a bit longer :-)

**Comment by kangbeicai** [comment]
(3) really doesn't matter , this is the case:
the files are serverl flac files and one cue file.  It has already been splited to the subtracks in the original directory.
like this:
![微信截图_20210204223729](https://user-images.githubusercontent.com/57837294/106907854-aa20b880-6739-11eb-8c6d-e61d9fbc8d87.png)
but because of (1) (**has been already solved** ),  the flac files are missing in the mount point.
 
I metioned it just because that for some music software like foobar, open cue file can load other flac files. Of course , I know such a case is not the job of trackfs. And as long as the files without cue files can be mounted normally, the results are the same. So I said it doesnt matter.

**Comment by andresch** [comment]
I've looked into (2):

The chardet library that I use to detect and read cuesheets in arbitrary charcter encodings complains about a wrong encoding of the file:
`UnicodeDecodeError: 'gb2312' codec can't decode byte 0xa9 in position 123: illegal multibyte sequence`

But I could open the file in Notepad++ and after converting it to UTF-8 trackfs could process the file correctly.

I've tested your file with two online character encoding tools and both had problems too. Ultimately I can't tell you if this a bug in chardet or if the file is really not properly encoded (but Notepad++ somehow finds a way to deal with it), but I'm afraid there is nothing I can really do at this point in time.

I'm really no expert on character sets, but maybe you find some tool that can convert the cue-sheets in your music-library from simplified chinese (gb2312) to utf8? If it's only an handful of files that have problems, notepad++ might allow you to do it manually. 

All I can do for now is not to fail for the whole directory in case of parsing errors and instead ignore the cue-sheet and expose the album file as regular file. Will create a ticket for this.

I'm closing this very long ticket that mixes up way too many topics. Don't hesitate to create new tickets if you run into any problems.
